File: typescript/LWTSD/ResourceTypes/Resource.ts
var xml = require('../../node-xmpp/packages/xml');

import {LWTSD} from '../index';
import {SimplifiedType} from '../SimplifiedType';
import {ResourceDescription} from '../ResourceDescription'
import { ResourceRestriction } from '../ResourceRestriction';

export abstract class Resource
{
    public Path : string;
    public ModifiedAt : Date;

    public Description : string;
    public Displayname : string;
    public Unit : string;
    public SupportsRead : boolean;
    public SupportsWrite : boolean;

    protected InnerMinExclusive : object;
    protected InnerMaxExclusive: object;
    protected Length : number;

    public readonly TotalPointsForWindow : number = 1;

    protected GetTimestampPoint(SimpleValue : string) : any
    {
        let RVal = xml("point",{ /*"xmlns" : LWTSD.name*/}, SimpleValue);
        RVal.attrs.timestamp = this.ModifiedAt.toISOString();

        return RVal;
    }

    protected GetDescriptionFromSimpleType(SimpleType: SimplifiedType) : ResourceDescription
    {
        let RVal = new ResourceDescription();
    
        RVal.Description = this.Description;
        RVal.DisplayName = this.Displayname;
        RVal.Path = this.Path;
        RVal.SimpleType = SimpleType;
        RVal.SupportsRead = this.SupportsRead;
        RVal.SupportsWrite = this.SupportsWrite;
        RVal.Restrictions = new ResourceRestriction();
        RVal.Restrictions.MinExclusive = this.InnerMinExclusive;
        RVal.Restrictions.MaxExclusive = this.InnerMaxExclusive;
        RVal.Restrictions.Length = this.Length;

        return RVal;
    }
    
    public abstract GetDescription() : ResourceDescription;
    public abstract LoadFromWrite(WriteElement : any) : void;
    public abstract GetWriteValue() : string;

    public constructor(Element? : any)
    {
        if(Element)
        {
            this.TotalPointsForWindow = parseInt(Element.attrs.totalpoints);
            this.Path = Element.attrs.path;
        }
    }

}

File: NET/LWTSD/ResourceTypes/Resource.cs
using System;
using System.Xml;
using System.Xml.Linq;

namespace LWTSD.ResourceTypes
{
	public abstract class Resource
	{
		public ResourcePath Path;
		public DateTime ModifiedAt;

		public string Description;
		public string Displayname;
		public string Unit;
		public bool SupportsRead;
		public bool SupportsWrite;

		protected object InnerMinExclusive;
		protected object InnerMaxExclusive;
		protected int? Length;

		public readonly int TotalPointsForWindow = 1;

		public virtual XElement GetPoint()
		{
			XElement RVal = new XElement(LWTSD.Namespace + "point");
			RVal.SetAttributeValue("timestamp",
									XmlConvert.ToString(ModifiedAt, XmlDateTimeSerializationMode.Utc));

			return RVal;

		}

		protected ResourceDescription GetDescription(SimplifiedType SimpleType)
		{
			return new ResourceDescription()
			{
				Description = Description,
				DisplayName = Displayname,
				Path = Path,
				SimpleType = SimpleType,
				SupportsRead = SupportsRead,
				SupportsWrite = SupportsWrite,
				Restrictions = new ResourceDescription.Restriction()
				{
					MinExclusive = InnerMinExclusive,
					MaxExclusive = InnerMaxExclusive,
					Length = Length
				}
			};
		}
		public abstract ResourceDescription GetDescription();
		public abstract void LoadFromWrite(XElement WriteElement);
		public abstract string GetWriteValue();

		public Resource(XElement Element)
		{
			TotalPointsForWindow = XmlConvert.ToInt32(Element.Attribute("totalpoints").Value);
			Path = Element.Attribute("path").Value;
		}

		public Resource()
		{
		}
	}
}
